### PR TITLE
Support Jingle file uploads

### PR DESCRIPTION
**Pull request by Half-Shot**
This PR is quite large, and I suggest reviewing commit-by-commit. Really these changes should land as separate PRs, but we're against the clock on these changes.

Notably, this PR adds:
- Support for transferring files in a DM from a XMPP -> Matrix user (via Jingle)
- Basic support for contact lists / rosters (allowing XMPP users to add Matrix users to their rosters)
- Sending capabilities to other clients (to enable features like Jingle)
- Sending presence of Matrix users to other XMPP users.

This PR is still due a bunch of cleanup, and is mostly here as a working example.

### CODE REVIEW

File: src/AutoRegistration.ts [hunk 1]
```diff
@@ -106,7 +108,17 @@ export class AutoRegistration {
...
+            if (!domainParameter) {
+                throw Error('`domain` must be specified in autoregistration parameters when only using a localpart')
+            }
+            let regexStr = usernameFormat.replace(/[-\/\\^$*+?.()|[\]{}]/g, "\\$&");
+            regexStr = regexStr.replace("<T_LOCALPART>", "(.+)");
+            const match = new RegExp(regexStr).exec(username);
+            if (!match || match.length < 2) {
+                throw Error("String didn't match");
+            }
+            log.debug("Result:", match);
```
Comment (tadzik) [inline src/AutoRegistration.ts]: Looks like a leftover – if not, maybe it should say "result of autoregitstration blah blah".

File: src/AutoRegistration.ts [hunk 2]
```diff
@@ -106,7 +108,17 @@ export class AutoRegistration {
             log.debug("Result:", match);
             mxid = `@${match[1]}:${match[2]}`;
         } else if (hasLocalpart) {
-            throw Error("We don't support localpart only, yet.");
+            if (!domainParameter) {
+                throw Error('`domain` must be specified in autoregistration parameters when only using a localpart')
+            }
+            let regexStr = usernameFormat.replace(/[-\/\\^$*+?.()|[\]{}]/g, "\\$&");
```
Comment (tadzik) [inline src/AutoRegistration.ts]: Hmm, what is this doing? The more I read the more confusing it appaers :) Seems to swap all possible regex metacharacters for `\$&`, but why?

File: src/Program.ts [hunk 3]
```diff
@@ -224,72 +233,100 @@ class Program {
             registration: this.cli.getRegistrationFilePath(),
             ...storeParams,
         });
+        await this.bridge.initalise();
+
+        this.store = await initiateStore(this.config.datastore, this.bridge);
+        const ignoreIntegrity = process.env.BIFROST_INTEGRITY_WRITE;
+        await this.store.integrityCheck(
+            ignoreIntegrity === undefined || ignoreIntegrity !== "false");
```
Comment (tadzik) [inline src/Program.ts:241]: I'm not sure how these map to the intention of `canWrite`? If we ignore integrity we allow modifying the DB?

File: src/Program.ts [hunk 4]
```diff
@@ -224,72 +233,100 @@ class Program {
...
-        await this.waitForHomeserver();
-        await this.bridge.listen(port);
-        log.info("Started appservice listener on port", port);
-        await this.pingBridge();
-        await this.registerBot();
 
-        if (this.cfg.metrics.enabled) {
-            log.info("Enabling metrics");
-            Metrics.init(this.bridge);
+        if (purple instanceof XmppJsInstance) {
```
Comment (tadzik) [inline src/Program.ts]: I think I'd rather have `preStart` in IBifrostInstance (unused in Purple) than introduce a(nother) specialcase branch here.

File: src/Program.ts [hunk 5]
```diff
@@ -224,72 +233,100 @@ class Program {
...
+                        await this.store.getAllIMRoomsForAccount(ev.mxid, acct.protocol.id)
+                    ).map((r) => r.remote.get<string>('recipient'));
+                    await acct.setPresence(presence, allInterestedUsers);
+                } catch (ex) {
+                    log.warn(`Failed to set startup presence for ${ev.mxid}`, ex);
+                }
+            }
+        });
+        purple.on("account-connection-error", (ev: IAccountEvent) => {
+            log.warn(`${ev.account.protocol_id}://${ev.account.username} had a connection error`, ev);
```
Comment (tadzik) [inline src/Program.ts:299]: Could be worth logging to the Bifrost room (maybe as a TODO if too much of a hassle now).

File: src/xmppjs/Jingle.ts [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+        this.assertState(JingleFileTransferSessionState.NeedOpen);
+        // Open a request to the media repo
+        const searchParams = new URLSearchParams({
+            filename: this.file.name || "file.bin",
+        });
+        const url = new URL(`/_matrix/media/r0/upload?${searchParams}`, this.uploadOpts.homeserverUrl);
+        this.stream = http.request(url, {
+            headers: {
+                'Content-Type': this.file.mediaType || "application/octect-stream",
+                // TODO: This will break if we are not given the size.
```
Comment (tadzik) [inline src/xmppjs/Jingle.ts:158]: I'm wondering how could we gently handle it in a standard-compliant way – size is indeed optional, "SHOULD" be here, but we also should accept the lack of it. I guess we could buffer the smaller files (with configurable max size) and only upload them once we have the entire thing. This'll turn the whole state machine upside down though currently, so just a food for thought for later.

File: src/xmppjs/Jingle.ts [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    public onData(b64: string) {
+        this.assertState(JingleFileTransferSessionState.Writable);
+        const buffer = Buffer.from(b64, "base64");
+        this.stream.write(buffer);
+        this.hashSHA1.write(buffer);
+    }
+
+    public onClose() {
+        this.assertState(JingleFileTransferSessionState.Writable);
+        this.stream.end();
```
Comment (tadzik) [inline src/xmppjs/Jingle.ts:180]:
```suggestion
        this.state = JingleFileTransferSessionState.Closed;
        this.stream.end();
```

Comment (tadzik) [reply src/xmppjs/Jingle.ts:180]: I wish Github would do "normalize" this diff :| Maybe git itself will.

File: src/xmppjs/Jingle.ts [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+        sid: string,
+        sessionDescription: JingleSessionInitiateOpts,
+        private readonly uploadOpts: JingleUploadOpts,
+        private readonly onUploadResult: (err: Error|null, mxcUrl?: string, file?: JingleFile) => void) {
+        super(sid, sessionDescription)
+    }
+
+    private assertState(...expectedStates: JingleFileTransferSessionState[]) {
+        if (!expectedStates.includes(this.state)) {
+            throw Error(`Invalid session state: ${this.state}`);
```
Comment (tadzik) [inline src/xmppjs/Jingle.ts:81]: I think throwing errors across this class qualifies for setting the state to Error and reporting back to onUploadResult – it's unlikely that it'll end successfully if the other side is not following the protocol. Notably, we don't have any error handling on the other end.

Being clearly stateful this makes perfect sense as an object, but perhaps it's worth wrapping in some Promising (sorry) functional interface for the caller.

File: src/xmppjs/Jingle.ts [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+        }
+        let session: JingleSession;
+        switch (descriptionXmlns) {
+            case XMPPFeatures.JingleFileTransferV4:
+            case XMPPFeatures.JingleFileTransferV5:
+                session = new JingleFileTransferSession(sid, info, this.uploadOpts, (err, mxcUrl, file) => {
+                    if (err) {
+                        // TODO: Cancel / emit a error to the other user.
+                        log.warn(`There was an error handling a Jingle file transfer for ${sid}:`, err);
+                        return;
```
Comment (tadzik) [inline src/xmppjs/Jingle.ts:259]: Should probably throw here also.

File: src/xmppjs/Jingle.ts [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+        const session = [...this.sessions.values()].find((s) => s instanceof JingleFileTransferSession && s.transportSid === sid) as JingleFileTransferSession;
+        if (!session) {
+            // Not found
+            // TODO: Improve error handling
+            return this.xmpp.xmppSend(
+                new StzaIqResponse(to, from, id, "error")
+            );
+        }
+        if (openElement) {
+            log.info(`Got open for session`);
```
Comment (tadzik) [inline src/xmppjs/Jingle.ts]:
```suggestion
            log.debug(`Got open for session`);
```

File: src/xmppjs/Jingle.ts [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+            session.onOpen();
+        } else if (dataElement) {
+            log.info(`Got data for session`);
```
Comment (tadzik) [inline src/xmppjs/Jingle.ts]:
```suggestion
            log.debug(`Got data for session`);
```

File: src/xmppjs/Jingle.ts [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+            session.onData(dataElement.getText());
+        } else if (closeElement) {
+            log.info(`Got close for session`);
```
Comment (tadzik) [inline src/xmppjs/Jingle.ts]:
```suggestion
            log.debug(`Got close for session`);
```

File: src/xmppjs/Stanzas.ts [hunk 13]
```diff
@@ -87,12 +90,32 @@ export class StzaPresence extends StzaBase {
...
+}
+
+
+export class StzaPresenceAvailable extends StzaPresence {
+    constructor(from: string, to?: string, type?: string, private capsHash?: string, private status?: string) {
+        super(from, to, uuid(), type);
+        this.includeXContent = false;
+    }
+    get presenceContent() {
+        const hash = this.capsHash ? `<c xmlns='http://jabber.org/protocol/caps' hash='sha-1' node='${NODE_NAME}' ver='${encode(this.capsHash)}'/>` : "";
```
Comment (tadzik) [inline src/xmppjs/Stanzas.ts:110]: The amount of hand-written XML we do (in general, not just this PR) is a little unsettling. Surely we have something in the dependency stack that'll allow to do this correctly and at least semi-easily. A thing for another task and another time :)

**Comment by Neustradamus** [comment]
@Half-Shot: Have you progressed on it?

File: src/AutoRegistration.ts (hunk 2 above)
Comment (Half-Shot) [reply src/AutoRegistration.ts]: I have forgotten totally what this did :(

**Comment by Half-Shot** [comment]
(Merged, this works fairly well, and at this point I just want to get the damn thing merged)

**Pull request merged**